cmd/main.go

```diff
@@ -394,6 +468,116 @@ func listKustomizations(ctx context.Context, client *rest.RESTClient) ([]ksapi.K
 	return ks.Items, err
 }
 
+func getClustersClientsFromKustomizationsAndHelmReleases(ctx context.Context, gitopsClient dynamic.Interface, kustomizations []ksapi.Kustomization, helmReleases []helmapi.HelmRelease) (map[string]clusterClients, error) {
+	resourcesWithSecrets := map[string]*unstructured.Unstructured{}
+	clients := make(map[string]clusterClients)
+
+	for _, ks := range kustomizations {
+		ks := ks
+		if ks.Spec.KubeConfig != nil {
+			key := fmt.Sprintf("%s/%s", ks.Namespace, ks.Spec.KubeConfig.SecretRef.Name)
+			if _, ok := resourcesWithSecrets[key]; !ok {
+				ksu, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&ks)
+				if err != nil {
+					return nil, err
+				}
+				resourcesWithSecrets[key] = &unstructured.Unstructured{Object: ksu}
+			}
+		}
+	}
+
+	for _, hr := range helmReleases {
+		hr := hr
+		if hr.Spec.KubeConfig != nil {
+			key := fmt.Sprintf("%s/%s", hr.Namespace, hr.Spec.KubeConfig.SecretRef.Name)
+			if _, ok := resourcesWithSecrets[key]; !ok {
+				hru, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&hr)
+				if err != nil {
+					return nil, err
+				}
+				resourcesWithSecrets[key] = &unstructured.Unstructured{Object: hru}
+			}
+		}
+	}
+
+	for _, r := range resourcesWithSecrets {
+		clusterName, clusterClts, err := getClusterClientsFromConfig(ctx, gitopsClient, r.GetNamespace(), r.Object["spec"])
+		if err != nil {
+			return nil, err
+		}
+
+		clients[clusterName] = clusterClts
+	}
+
+	return clients, nil
+}
+
+func loadKubeconfigSecret(ctx context.Context, gitopsClient dynamic.Interface, namespace, name string) (*v1.Secret, error) {
+	var secret v1.Secret
+	element, err := gitopsClient.Resource(schema.GroupVersionResource{
+		Group:    "",
+		Version:  "v1",
+		Resource: "secrets",
+	}).Namespace(namespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(element.UnstructuredContent(), &secret)
+	if err != nil {
+		return nil, err
+	}
+
+	return &secret, nil
+}
+
+func getClusterClientsFromConfig(ctx context.Context, gitopsClient dynamic.Interface, namespace string, specStr interface{}) (string, clusterClients, error) {
+	spec := ksapi.KustomizationSpec{}
+	b, err := json.Marshal(specStr)
+	if err != nil {
+		return "", clusterClients{}, err
+	}
+
+	err = json.Unmarshal(b, &spec)
+	if err != nil {
+		return "", clusterClients{}, err
+	}
+
+	secret, err := loadKubeconfigSecret(ctx, gitopsClient, namespace, spec.KubeConfig.SecretRef.Name)
+	if err != nil {
+		return "", clusterClients{}, err
+	}
+
+	key := "value"
+	if spec.KubeConfig.SecretRef.Key != "" {
+		key = spec.KubeConfig.SecretRef.Key
+	}
```

Comment:
It should check both value and value.yaml if no key is set.
Afaict flux checks also for both and in the docs, value.yaml is documented.
(for example https://github.com/fluxcd/pkg/blob/ccc3bab767f72ff00f01f29eeeb8373655ac0c2f/runtime/client/impersonator.go#L186)